### PR TITLE
Download command

### PR DESCRIPTION
**Pull request by jhoxhaa**

### CODE REVIEW

**Review by lippserd — CHANGES_REQUESTED** [review]
Remove the `exit;` calls and do the following for the file creation:

```php
switch ($format) {
    case 'csv':
        $content = $report->toCsv();
        break;
    case 'json':
        $content = $report->toJson();
        break;
    case 'pdf':
        $content = Pdfexport::first()...;
        break;
    default:
        throw ...
}

$filename = getcwd() . "/$name.$format";
file_put_contents($filename, $content);
echo "$filename\n";
```

Note the difference between `dirname(__FILE__)` and `getcwd()`. Why should we prefer `getcwd()`?

File: application/clicommands/DownloadCommand.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                $content = $report->toJson();
+                break;
+            default:
+                throw new Exception('Format is invalid');
+        }
+
+        $filename = getcwd() . "/$name.$format";
+        file_put_contents($filename, $content);
+        echo "$filename\n";
+        exit;
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:50]: The `exit()` call here is not only superfluous, but also incorrect. The CLI command action is executed by a dispatcher, which executes code before and after the action is executed. With `exit()`, the dispatcher cannot execute any code after the action. Please try the download command with the `--benchmark` option with and without the `exit()` call and compare the results.

File: application/clicommands/DownloadCommand.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                $content = Pdfexport::first()->htmlToPdf($report->toPdf());
+                break;
+            case 'csv':
+                $content = $report->toCsv();
+                break;
+            case 'json':
+                $content = $report->toJson();
+                break;
+            default:
+                throw new Exception('Format is invalid');
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:44]: `throw new InvalidArgumentException(sprintf('Format %s is not supported', $format));`

Unsupported is friendlier than invalid, and it shows exactly what is wrong.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: application/clicommands/DownloadCommand.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
+<?php
+
```
Comment (yhabteab) [inline application/clicommands/DownloadCommand.php:2]: Missing License header

File: application/clicommands/DownloadCommand.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                break;
+            case 'json':
+                $content = $report->toJson();
+                break;
+            default:
+                throw new InvalidArgumentException(sprintf('Format %s is not supported', $format));
+        }
+
+        $storage = new LocalFileStorage(Icinga::app()->getStorageDir(
+            join(DIRECTORY_SEPARATOR, ['modules', 'reporting', 'reports'])
```
Comment (yhabteab) [inline application/clicommands/DownloadCommand.php:62]: Please use `implode()` instead of `join()`.

File: library/Reporting/Web/Forms/ReportForm.php [hunk 5]
```diff
@@ -34,8 +35,22 @@ protected function assemble()
         $this->setDefaultElementDecorator(new CompatDecorator());
 
         $this->addElement('text', 'name', [
-            'required' => true,
-            'label'    => 'Name'
+            'required'  => true,
+            'label'     => 'Name',
+            'validators' => [
+                'Callback' => function ($value, $validator) {
+                    /** @var CallbackValidator $validator */
```
Comment (yhabteab) [inline library/Reporting/Web/Forms/ReportForm.php:42]:
```suggestion
                'Callback' => function ($value, CallbackValidator $validator) {
```

File: library/Reporting/Web/Forms/ReportForm.php [hunk 6]
```diff
@@ -34,8 +35,22 @@ protected function assemble()
...
+                    if ($value[0] === '.') {
+                        $validator->addMessage(
+                            $this->translate('Leading dots qare not allowed in the report name')
```
Comment (yhabteab) [inline library/Reporting/Web/Forms/ReportForm.php:45]: **... dots qare** -> `... dots are`

**Review by yhabteab — CHANGES_REQUESTED** [review]

**Comment by yhabteab** [comment]
Don't forget also to rebase the branch.

File: application/clicommands/DownloadCommand.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
+<?php
+
+// Icinga Reporting | (c) 2018 Icinga GmbH | GPLv2
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:3]: Please replace 2018 with 2023 😆

File: application/clicommands/DownloadCommand.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                $content = $report->toCsv();
+                break;
+            case 'json':
+                $content = $report->toJson();
+                break;
+            default:
+                throw new InvalidArgumentException(sprintf('Format %s is not supported', $format));
+        }
+
+        $storage = new LocalFileStorage(Icinga::app()->getStorageDir(
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:63]: Why are we using the storage now? Have I requested that change?

File: library/Reporting/Web/Forms/ReportForm.php [hunk 9]
```diff
@@ -49,7 +50,20 @@ protected function assemble()
             'description' => $this->translate(
                 'A unique name of this report. It is used when exporting to pdf, json or csv format'
                 . ' and also when listing the reports in the cli'
-            )
+            ),
+            'validators' => [
+                'Callback' => function ($value, CallbackValidator $validator) {
+                    if ($value !== null && $value[0] === '.') {
```
Comment (lippserd) [inline library/Reporting/Web/Forms/ReportForm.php:56]: I think I suggested something like this but it's not enough to protect from path traversal. `/../../owned.pdf` would be valid but that's not something we want. Single dots are no harm even at the beginning as would be just a hidden file. Please research "path traversal" a bit and suggest how to prevent it here.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: application/clicommands/DownloadCommand.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    public function defaultAction()
+    {
+        $id = $this->params->getStandalone();
+        if ($id === null) {
+            $this->fail($this->translate('Argument id is mandatory'));
+        }
+
+        $report = Report::fromDb($id);
+        $format = strtolower($this->params->get('format', 'pdf'));
+        $name = sprintf(
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:49]: This is only needed if `output` is not specified.

File: library/Reporting/Web/Forms/ReportForm.php [hunk 11]
```diff
@@ -49,7 +50,20 @@ protected function assemble()
             'description' => $this->translate(
                 'A unique name of this report. It is used when exporting to pdf, json or csv format'
                 . ' and also when listing the reports in the cli'
-            )
+            ),
+            'validators' => [
```
Comment (lippserd) [inline library/Reporting/Web/Forms/ReportForm.php:54]: The added validator should be made in a separate commit.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: application/clicommands/DownloadCommand.php [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $id = $this->params->getStandalone();
+        if ($id === null) {
+            $this->fail($this->translate('Argument id is mandatory'));
+        }
+
+        $output = $this->params->get('output');
+        if ($output === null) {
+            $report = Report::fromDb($id);
+            $format = strtolower($this->params->get('format', 'pdf'));
```
Comment (lippserd) [inline application/clicommands/DownloadCommand.php:50]: Handling `$format` and `$content` is still required to run before the `if`. Now you don't have content if you specify `output`.

**Review by lippserd — CHANGES_REQUESTED** [review]

**Review by lippserd — APPROVED** [review]

**Review by yhabteab — APPROVED** [review]
